pkg/scheduler/scheduler.go

```diff
@@ -424,9 +439,9 @@ func (s *Scheduler) getAssignments(log logr.Logger, wl *workload.Info, snap *cac
 	}
 
 	if arm == flavorassigner.Preempt {
-		faPreemptionTargets := s.preemptor.GetTargets(log, *wl, fullAssignment, snap)
+		preemptionAssignment, faPreemptionTargets := s.preemptor.GetTargets(log, *wl, fullAssignment, snap)
 		if len(faPreemptionTargets) > 0 {
```

Comment:
It seems this will update `fullAssignment` as well, right? Because preemptor does not clone `fullAssignment` and just update it.
So, even if `len(faPreemptionTargets)` is `0`, the `fullAssignment` looks like updated in `GetTargets()` function.

I meant even if there are no preemption targets, the assignment will be updated.
Is this intended behavior?

Reply:
Good point, actually the `fullAssignment` is modified by GetTargets and the new value is returned. So, we have two options:
1. don't return the new value and rename the function to something reflecting it may mutate assignment, like "Preempt"
2. clone the assignment and return the updated version

I would be leaning to 2, but it is more involving as we need to implement Clone for Assignment and its substructure. WDYT?

> So, even if len(faPreemptionTargets) is 0, the fullAssignment looks like updated in GetTargets() function.

It should not be updated, because if len(targets) is zero we wouldn't get successful tasResult, but we can prevent that explicitly in code.

Reply:
Ok, I took the "in the middle" approach. I no longer mutate the assignment in GetTargets, but outside in the getAssignments function, which makes sense IMO. The GetTargets just returns extra the tasResult. 

WDYT?

Reply:
I have a better idea - just return the targets and re-compute the assignment. This will also eliminate other comments. The drawback is slightly more computation, but I think we can optimize when needed. On it.

Reply:
Done, PTAL https://github.com/kubernetes-sigs/kueue/commit/f1a3f568c6a062fc5c4491e918957574fb032fa6